### PR TITLE
Default Cost Model

### PR DESCRIPTION
**Pull request by vivsy**
This is the default cost model ported over from Peloton and modified to use the current statistics storage code.

### CODE REVIEW

File: src/include/optimizer/physical_operators.h [hunk 1]
```diff
@@ -157,8 +158,8 @@ class IndexScan : public OperatorNode<IndexScan> {
    * @return an IndexScan operator
    */
   static Operator Make(catalog::db_oid_t database_oid, catalog::namespace_oid_t namespace_oid,
-                       catalog::index_oid_t index_oid, std::vector<AnnotatedExpression> &&predicates,
-                       std::string table_alias, bool is_for_update,
+                       catalog::index_oid_t index_oid, catalog::table_oid_t table_oid,
```
Comment (apavlo) [inline src/include/optimizer/physical_operators.h]: Since the table "owns" the index, the `table_oid_t` should come before `index_oid_t`.

Comment (apavlo) [inline src/include/optimizer/physical_operators.h]: Can you explain why you needed to add the `table_oid_t` here?

File: src/include/optimizer/cost_model/default_cost_model.h [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  }
+
+  /**
+   * Seq scan operator to visit
+   * @param op operator
+   */
+  void Visit(const SeqScan *op) override {
+    auto table_stats = stats_storage_->GetTableStats(op->GetDatabaseOID(), op->GetTableOID());
+    if (table_stats->GetColumnCount() == 0) {
+      output_cost_ = 1.f;
```
Comment (apavlo) [inline src/include/optimizer/cost_model/default_cost_model.h]: Magic constant.

File: src/include/optimizer/cost_model/default_cost_model.h [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+   * Index scan operator to visit
+   * @param op operator
+   */
+  void Visit(UNUSED_ATTRIBUTE const IndexScan *op) override {
+    auto table_stats = stats_storage_->GetTableStats(op->GetDatabaseOID(), op->GetTableOID());
+    if (table_stats->GetColumnCount() == 0 || table_stats->GetNumRows() == 0) {
+      output_cost_ = 0.f;
+      return;
+    }
+    output_cost_ = std::log2(table_stats->GetNumRows()) * DEFAULT_INDEX_TUPLE_COST +
```
[the remark this answers is not in the document]
Comment (vivsy) [reply src/include/optimizer/cost_model/default_cost_model.h]: They're defined at the bottom of the file (lines 247-255)

File: src/include/optimizer/physical_operators.h (hunk 1 above)
Comment (vivsy) [reply src/include/optimizer/physical_operators.h]: The stats storage object needs the table oid of the operator to fetch its table stats for evaluation in the cost model.

File: src/include/optimizer/physical_operators.h [hunk 4]
```diff
@@ -182,7 +181,7 @@ class IndexScan : public OperatorNode<IndexScan> {
   const catalog::namespace_oid_t &GetNamespaceOID() const { return namespace_oid_; }
 
   /**
-   * @return the OID of the table
+   * @return the OID of the index
```
Comment (apavlo) [inline src/include/optimizer/physical_operators.h]: Index or Table?

File: test/optimizer/default_cost_model_test.cpp [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+#include "optimizer/physical_operators.h"
+#include "parser/expression/abstract_expression.h"
+#include "parser/expression/constant_value_expression.h"
+#include "type/transient_value.h"
+#include "type/transient_value_factory.h"
+
+#include "test_util/test_harness.h"
+
+namespace terrier::optimizer {
+class DefaultCostModelTests : public TerrierTest {
```
Comment (apavlo) [inline test/optimizer/default_cost_model_test.cpp]: This is an impressive amount of code. Can you please add some comments in it to explain how it works and what each test is doing? This is will make it easier for new people to maintain/extend after you graduate.

**Comment by apavlo** [comment]
Supplanted by #870

**Pull request closed without merging**